File: generate_signal.js
/**
 * シグナル生成スクリプト
 * Usage: node generate_signal.js [--window 60] [--lambda 0.9] [--quantile 0.4]
 */

'use strict';

const fs = require('fs');
const path = require('path');

// ライブラリ
const { createLogger } = require('./lib/logger');
const { config } = require('./lib/config');
const { LeadLagSignal } = require('./lib/pca');
const { buildPortfolio, computePerformanceMetrics } = require('./lib/portfolio');
const { correlationMatrixSample } = require('./lib/math');
const { loadCSV, buildPaperAlignedReturnRows } = require('./lib/data');

const logger = createLogger('SignalGenerator');

// 米国セクター ETF
const US_ETF_TICKERS = [
  'XLB', 'XLC', 'XLE', 'XLF', 'XLI', 'XLK', 'XLP',
  'XLRE', 'XLU', 'XLV', 'XLY'
];

// 日本セクター ETF
const JP_ETF_TICKERS = [
  '1617.T', '1618.T', '1619.T', '1620.T', '1621.T', '1622.T', '1623.T',
  '1624.T', '1625.T', '1626.T', '1627.T', '1628.T', '1629.T', '1630.T',
  '1631.T', '1632.T', '1633.T'
];

const JP_ETF_NAMES = {
  '1617.T': '食品', '1618.T': 'エネルギー資源', '1619.T': '建設・資材',
  '1620.T': '素材・化学', '1621.T': '医薬品', '1622.T': '自動車・輸送機',
  '1623.T': '鉄鋼・非鉄', '1624.T': '機械', '1625.T': '電機・精密',
  '1626.T': '情報通信', '1627.T': '電力・ガス', '1628.T': '運輸・物流',
  '1629.T': '商社・卸売', '1630.T': '小売', '1631.T': '銀行',
  '1632.T': '証券・商品', '1633.T': '保険'
};

/**
 * Yahoo Financeからデータを取得
 */
async function fetchData(ticker, days = 200) {
  if (config.data.mode === 'csv') {
    const filePath = path.join(path.resolve(config.data.dataDir), `${ticker}.csv`);
    if (!fs.existsSync(filePath)) {
      logger.error(`CSV not found: ${filePath}`);
      return [];
    }
    try {
      const rows = loadCSV(filePath).map(row => ({
        date: String(row.Date || row.date || '').split('T')[0],
        open: Number(row.Open ?? row.open),
        high: Number(row.High ?? row.high),
        low: Number(row.Low ?? row.low),
        close: Number(row.Close ?? row.close),
        volume: Number(row.Volume ?? row.volume ?? 0)
      })).filter(r => r.date && Number.isFinite(r.close) && r.close > 0);
      return days > 0 && rows.length > days ? rows.slice(-days) : rows;
    } catch (error) {
      logger.error(`Failed to load ${ticker}`, { error: error.message });
      return [];
    }
  }

  try {
    const YahooFinance = require('yahoo-finance2').default;
    const yahooFinance = new YahooFinance();

    const endDate = new Date();
    const startDate = new Date();
    startDate.setDate(startDate.getDate() - days);

    const result = await yahooFinance.chart(ticker, {
      period1: startDate.toISOString().split('T')[0],
      period2: endDate.toISOString().split('T')[0],
      interval: '1d'
    });

    return result.quotes
      .filter(q => q.close !== null && q.close > 0)
      .map(q => ({
        date: q.date.toISOString().split('T')[0],
        open: q.open,
        high: q.high,
        low: q.low,
        close: q.close,
        volume: q.volume
      }));
  } catch (error) {
    logger.error(`Failed to fetch ${ticker}`, { error: error.message });
    return [];
  }
}

/**
 * リターンを計算
 */
function computeReturns(ohlc, type = 'cc') {
  if (!ohlc || ohlc.length < 2) return [];

  if (type === 'cc') {
    const returns = [];
    let prev = null;
    for (const r of ohlc) {
      if (prev !== null) {
        returns.push({ date: r.date, return: (r.close - prev) / prev });
      }
      prev = r.close;
    }
    return returns;
  } else {
    return ohlc
      .filter(r => r.open > 0)
      .map(r => ({ date: r.date, return: (r.close - r.open) / r.open }));
  }
}

/**
 * コマンドライン引数を解析
 */
function parseArgs() {
  const args = process.argv.slice(2);
  const options = {
    windowLength: config.backtest.windowLength,
    lambdaReg: config.backtest.lambdaReg,
    quantile: config.backtest.quantile,
    save: true
  };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--window':
      case '-w':
        options.windowLength = parseInt(args[++i]);
        break;
      case '--lambda':
      case '-l':
        options.lambdaReg = parseFloat(args[++i]);
        break;
      case '--quantile':
      case '-q':
        options.quantile = parseFloat(args[++i]);
        break;
      case '--no-save':
        options.save = false;
        break;
      case '--help':
      case '-h':
        console.log('Usage: node generate_signal.js [options]');
        console.log('Options:');
        console.log('  --window, -w <n>    Window length (default: 60)');
        console.log('  --lambda, -l <n>    Lambda regularization (default: 0.9)');
        console.log('  --quantile, -q <n>  Quantile (default: 0.4)');
        console.log('  --no-save           Do not save results');
        console.log('  --help, -h          Show this help');
        process.exit(0);
    }
  }

  return options;
}

/**
 * メイン処理
 */
async function main() {
  const options = parseArgs();

  logger.info('Signal generation started', options);

  // データ取得
  console.log('\n📡 Fetching data from Yahoo Finance...');

  const usData = {};
  for (const ticker of US_ETF_TICKERS) {
    process.stdout.write(`  ${ticker}... `);
    usData[ticker] = await fetchData(ticker, options.windowLength + 50);
    console.log(`${usData[ticker].length} days`);
  }

  const jpData = {};
  for (const ticker of JP_ETF_TICKERS) {
    process.stdout.write(`  ${ticker}... `);
    jpData[ticker] = await fetchData(ticker, options.windowLength + 50);
    console.log(`${jpData[ticker].length} days`);
  }

  const usCC = {};
  const jpCC = {};
  const jpOC = {};

  for (const t of US_ETF_TICKERS) {
    usCC[t] = computeReturns(usData[t], 'cc');
  }
  for (const t of JP_ETF_TICKERS) {
    jpCC[t] = computeReturns(jpData[t], 'cc');
    jpOC[t] = computeReturns(jpData[t], 'oc');
  }

  const usMap = new Map();
  for (const t of US_ETF_TICKERS) {
    for (const r of usCC[t]) {
      if (!usMap.has(r.date)) usMap.set(r.date, {});
      usMap.get(r.date)[t] = r.return;
    }
  }
  const jpCCMap = new Map();
  const jpOCMap = new Map();
  for (const t of JP_ETF_TICKERS) {
    for (const r of jpCC[t]) {
      if (!jpCCMap.has(r.date)) jpCCMap.set(r.date, {});
      jpCCMap.get(r.date)[t] = r.return;
    }
    for (const r of jpOC[t]) {
      if (!jpOCMap.has(r.date)) jpOCMap.set(r.date, {});
      jpOCMap.get(r.date)[t] = r.return;
    }
  }

  const { retUs, retJp } = buildPaperAlignedReturnRows(
    usMap,
    jpCCMap,
    jpOCMap,
    US_ETF_TICKERS,
    JP_ETF_TICKERS,
    config.backtest.jpWindowReturn
  );

  console.log(`\n📊 Data prepared: ${retUs.length} trading days`);

  if (retUs.length < options.windowLength) {
    logger.error('Insufficient data');
    console.error('Error: Insufficient data for window length');
    process.exit(1);
  }

  const combined = retUs.map((r, i) => [...r.values, ...retJp[i].values]);
  const CFull = correlationMatrixSample(combined);

  const signalGen = new LeadLagSignal({
    lambdaReg: options.lambdaReg,
    nFactors: config.backtest.nFactors,
    orderedSectorKeys: config.pca.orderedSectorKeys
  });

  const retUsWin = retUs.slice(-options.windowLength).map(r => r.values);
  const retJpWin = retJp.slice(-options.windowLength).map(r => r.values);
  const retUsLatest = retUs[retUs.length - 1].values;

  const signal = signalGen.computeSignal(
    retUsWin,
    retJpWin,
    retUsLatest,
    config.sectorLabels,
    CFull
  );

  // ランキング作成
  const signals = JP_ETF_TICKERS.map((ticker, i) => ({
    ticker,
    name: JP_ETF_NAMES[ticker] || ticker,
    sector: config.sectorLabels[`JP_${ticker}`] || 'unknown',
    signal: signal[i],
    rank: 0
  })).sort((a, b) => b.signal - a.signal);

  signals.forEach((s, i) => s.rank = i + 1);

  // 買い/売り候補
  const buyCount = Math.max(1, Math.floor(JP_ETF_TICKERS.length * options.quantile));
  const buyCandidates = signals.slice(0, buyCount);
  const sellCandidates = signals.slice(-buyCount);

  // 結果表示
  console.log('\n' + '='.repeat(60));
  console.log('📈 買い銘柄（ロング）');
  console.log('='.repeat(60));
  console.log('Rank  Ticker    業種           シグナル値');
  console.log('-'.repeat(60));

  buyCandidates.forEach((s, i) => {
    console.log(
      `${String(i + 1).padStart(4)}  ${s.ticker.padEnd(8)} ${s.name.padEnd(12)} ${(s.signal * 1000).toFixed(2)}`
    );
  });

  console.log('\n' + '='.repeat(60));
  console.log('📉 売り銘柄（ショート）');
  console.log('='.repeat(60));
  console.log('Rank  Ticker    業種           シグナル値');
  console.log('-'.repeat(60));

  sellCandidates.slice().reverse().forEach((s, i) => {
    console.log(
      `${String(buyCount - i).padStart(4)}  ${s.ticker.padEnd(8)} ${s.name.padEnd(12)} ${(s.signal * 1000).toFixed(2)}`
    );
  });

  // シグナル統計
  const meanSignal = signal.reduce((a, b) => a + b, 0) / signal.length;
  const stdSignal = Math.sqrt(signal.reduce((sq, x) => sq + Math.pow(x - meanSignal, 2), 0) / signal.length);

  console.log('\n' + '='.repeat(60));
  console.log('📊 シグナル統計');
  console.log('='.repeat(60));
  console.log(`  平均: ${(meanSignal * 1000).toFixed(4)}`);
  console.log(`  標準偏差: ${(stdSignal * 1000).toFixed(4)}`);
  console.log(`  最大: ${(Math.max(...signal) * 1000).toFixed(4)}`);
  console.log(`  最小: ${(Math.min(...signal) * 1000).toFixed(4)}`);

  // 結果保存
  if (options.save) {
    const outputDir = config.data.outputDir;
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    const result = {
      timestamp: new Date().toISOString(),
      config: options,
      signals,
      buyCandidates,
      sellCandidates,
      statistics: { meanSignal, stdSignal }
    };

    const jsonPath = path.join(outputDir, 'signal.json');
    fs.writeFileSync(jsonPath, JSON.stringify(result, null, 2));

    const csvPath = path.join(outputDir, 'signal_long.csv');
    const csvContent = 'Rank,Ticker,Name,Sector,Signal\n' +
      buyCandidates.map((s, i) =>
        `${i + 1},${s.ticker},${s.name},${s.sector},${s.signal}`
      ).join('\n');
    fs.writeFileSync(csvPath, csvContent);

    console.log(`\n💾 Results saved:`);
    console.log(`  - ${jsonPath}`);
    console.log(`  - ${csvPath}`);
  }

  logger.info('Signal generation completed');
}

main().catch(error => {
  logger.error('Signal generation failed', { error: error.message, stack: error.stack });
  console.error('Error:', error.message);
  process.exit(1);
});

File: streamlit_app/generate_signal.py
"""
日米業種リードラグ戦略 - シグナル生成（Python 版）

Streamlit Cloud 環境用
"""
import json
import numpy as np
import pandas as pd
from pathlib import Path
from datetime import datetime, timedelta
import requests

# yfinance の使用を試みる（なければ requests で直接取得）
USE_YFINANCE = False
try:
    import yfinance as yf
    USE_YFINANCE = True
    print("✅ yfinance を使用します")
except ImportError:
    print("ℹ️ yfinance がないため、requests で直接データ取得します")


def fetch_etf_data(tickers: list[str], period: str = "1y") -> dict[str, pd.DataFrame]:
    """ETF データを取得"""
    data = {}
    
    if USE_YFINANCE:
        # yfinance を使用
        for ticker in tickers:
            try:
                df = yf.download(ticker, period=period, progress=False)
                if len(df) > 0:
                    data[ticker] = df
            except Exception as e:
                print(f"Error fetching {ticker}: {e}")
    else:
        # requests で直接 Yahoo Finance から取得
        period_map = {"1y": 365, "6mo": 180, "3mo": 90}
        days = period_map.get(period, 365)
        end_date = datetime.now()
        start_date = end_date - timedelta(days=days)
        
        for ticker in tickers:
            try:
                url = f"https://query1.finance.yahoo.com/v8/finance/chart/{ticker}"
                params = {
                    "period1": int(start_date.timestamp()),
                    "period2": int(end_date.timestamp()),
                    "interval": "1d"
                }
                response = requests.get(url, params=params, timeout=10)
                response.raise_for_status()
                result = response.json()
                
                if result.get("chart", {}).get("result"):
                    chart_data = result["chart"]["result"][0]
                    timestamps = chart_data["timestamp"]
                    quotes = chart_data["indicators"]["quote"][0]
                    
                    df = pd.DataFrame({
                        "Date": pd.to_datetime(timestamps, unit="s"),
                        "Close": quotes["close"]
                    })
                    df = df.set_index("Date")
                    df = df.dropna()
                    
                    if len(df) > 0:
                        data[ticker] = df
            except Exception as e:
                print(f"Error fetching {ticker} via requests: {e}")
    
    return data


def calculate_returns(prices: pd.Series) -> pd.Series:
    """対数リターン計算"""
    return np.log(prices / prices.shift(1)).dropna()


def build_return_matrix(data: dict[str, pd.DataFrame]) -> np.ndarray:
    """リターン行列構築"""
    returns_list = []
    for ticker, df in data.items():
        close = df['Close'] if isinstance(df, pd.DataFrame) else df
        returns = calculate_returns(close)
        returns_list.append(returns.values)
    
    # 日付を揃えて行列化
    max_len = min(len(r) for r in returns_list)
    returns_matrix = np.array([r[-max_len:] for r in returns_list])
    return returns_matrix.T  # 日数 × 銘柄数


def pca_subspace_signal(returns: np.ndarray, n_factors: int = 3, 
                        lambda_reg: float = 0.9, window: int = 60) -> np.ndarray:
    """
    部分空間正則化 PCA シグナル（簡易版）
    
    注意：これは簡易実装です。完全な実装は lib/pca/subspace.js を参照
    """
    if len(returns) < window:
        # データが不足している場合は単純モメンタム
        momentum = returns[-1] - returns[-min(window, len(returns))]
        return momentum
    
    # 直近 window 日のデータを使用
    window_returns = returns[-window:]
    
    # 相関行列計算
    corr_matrix = np.corrcoef(window_returns.T)
    
    # 固有値分解（簡易版）
    try:
        eigenvalues, eigenvectors = np.linalg.eigh(corr_matrix)
        # 上位 n_factors 個の固有値・固有ベクトルを取得
        idx = np.argsort(eigenvalues)[::-1][:n_factors]
        eigenvalues = eigenvalues[idx]
        eigenvectors = eigenvectors[:, idx]
    except Exception:
        # 固有値分解失敗時は単純平均
        eigenvectors = np.ones((corr_matrix.shape[0], n_factors)) / np.sqrt(n_factors)
    
    # 最新リターンを射影
    latest_returns = returns[-1]
    factor_scores = latest_returns @ eigenvectors
    
    # シグナル復元
    signal = factor_scores @ eigenvectors.T
    
    return signal.sum(axis=1)


def generate_signal() -> dict:
    """シグナル生成"""
    # 銘柄リスト
    US_ETF_TICKERS = ['XLB', 'XLC', 'XLE', 'XLF', 'XLI', 'XLK', 
                      'XLP', 'XLRE', 'XLU', 'XLV', 'XLY']
    JP_ETF_TICKERS = ['1617.T', '1618.T', '1619.T', '1620.T', '1621.T',
                      '1622.T', '1623.T', '1624.T', '1625.T', '1626.T',
                      '1627.T', '1628.T', '1629.T', '1630.T', '1631.T',
                      '1632.T', '1633.T']
    
    # セクターラベル
    SECTOR_LABELS = {
        '1617.T': '食品', '1618.T': 'エネルギー資源', '1619.T': '化学',
        '1620.T': '素材・化学', '1621.T': '医薬品', '1622.T': '鉄鋼',
        '1623.T': '鉄鋼・非鉄', '1624.T': '機械', '1625.T': '電機・精密',
        '1626.T': '自動車', '1627.T': '電力・ガス', '1628.T': '運輸',
        '1629.T': '商社・卸売', '1630.T': '小売', '1631.T': '銀行',
        '1632.T': '証券・商品', '1633.T': '保険'
    }
    
    print("📡 データ取得中...")
    
    # データ取得
    all_tickers = US_ETF_TICKERS + JP_ETF_TICKERS

    if USE_YFINANCE:
        data = fetch_etf_data(all_tickers, period="1y")
    else:
        # requests で直接取得
        print("📡 requests でデータ取得中...")
        data = fetch_etf_data(all_tickers, period="1y")
    
    if not data:
        return {"error": "データ取得に失敗しました"}
    
    print(f"📊 {len(list(data.values())[0])} days of data retrieved")
    
    # リターン計算
    print("📈 リターン計算中...")
    returns_dict = {}
    for ticker, df in data.items():
        close = df['Close'] if isinstance(df, pd.DataFrame) else df
        returns_dict[ticker] = calculate_returns(close)
    
    # 共通の期間に揃える
    min_len = min(len(r) for r in returns_dict.values())
    returns_array = np.array([list(r.values[-min_len:]) for r in returns_dict.values()]).T
    
    # シグナル生成
    print("📉 シグナル生成中...")
    signals = pca_subspace_signal(returns_array, n_factors=3, lambda_reg=0.9, window=60)
    
    # 結果構築
    results = []
    for i, ticker in enumerate(returns_dict.keys()):
        if ticker in JP_ETF_TICKERS:
            # numpy 配列を Python スカラーに変換
            signal_val = signals[i]
            if hasattr(signal_val, 'item'):
                signal_val = signal_val.item()
            results.append({
                'ticker': ticker,
                'sector': SECTOR_LABELS.get(ticker, 'Unknown'),
                'signal': float(signal_val),
                'region': 'JP'
            })
    
    # シグナルでソート
    results.sort(key=lambda x: x['signal'], reverse=True)
    
    # 統計量
    signal_values = [r['signal'] for r in results]
    
    return {
        'timestamp': datetime.now().isoformat(),
        'signals': results,
        'config': {
            'windowLength': 60,
            'lambdaReg': 0.9,
            'quantile': 0.4,
            'nFactors': 3
        },
        'statistics': {
            'mean': float(np.mean(signal_values)),
            'std': float(np.std(signal_values)),
            'max': float(np.max(signal_values)),
            'min': float(np.min(signal_values))
        }
    }


def main():
    """メイン処理"""
    print("=" * 60)
    print("📊 日米業種リードラグ戦略 - シグナル生成（Python 版）")
    print("=" * 60)
    
    result = generate_signal()
    
    if 'error' in result:
        print(f"❌ エラー：{result['error']}")
        return
    
    # 結果表示
    print("\n" + "=" * 60)
    print("📈 買い銘柄（ロング）")
    print("=" * 60)
    
    long_candidates = [r for r in result['signals'] if r['signal'] > 0][:6]
    for rank, r in enumerate(long_candidates, 1):
        print(f"Rank {rank:2d}  {r['ticker']:8s}  {r['sector']:12s}  {r['signal']:8.2f}")
    
    print("\n" + "=" * 60)
    print("📉 売り銘柄（ショート）")
    print("=" * 60)
    
    short_candidates = [r for r in result['signals'] if r['signal'] < 0][:6]
    short_candidates.sort(key=lambda x: x['signal'])
    for rank, r in enumerate(short_candidates, 1):
        print(f"Rank {rank:2d}  {r['ticker']:8s}  {r['sector']:12s}  {r['signal']:8.2f}")
    
    print("\n" + "=" * 60)
    print("📊 シグナル統計")
    print("=" * 60)
    stats = result['statistics']
    print(f"  平均：{stats['mean']:.4f}")
    print(f"  標準偏差：{stats['std']:.4f}")
    print(f"  最大：{stats['max']:.4f}")
    print(f"  最小：{stats['min']:.4f}")
    
    # 結果を保存
    output_dir = Path(__file__).parent.parent / 'results'
    output_dir.mkdir(exist_ok=True)
    
    with open(output_dir / 'signal_python.json', 'w', encoding='utf-8') as f:
        json.dump(result, f, indent=2, ensure_ascii=False)
    
    print(f"\n💾 Results saved to: {output_dir / 'signal_python.json'}")
    print("\n✅ シグナル生成完了")


if __name__ == '__main__':
    main()
